src/coreclr/jit/importercalls.cpp

```diff
@@ -4042,57 +4036,26 @@ GenTree* Compiler::impSRCSUnsafeIntrinsic(NamedIntrinsic        intrinsic,
                 return nullptr;
             }
 
-            CorInfoType fromJitType = info.compCompHnd->asCorInfoType(fromTypeHnd);
-            var_types   fromType    = JITtype2varType(fromJitType);
+            ClassLayout* fromLayout = nullptr;
+            var_types    fromType   = TypeHandleToVarType(fromTypeHnd, &fromLayout);
 
-            CorInfoType toJitType = info.compCompHnd->asCorInfoType(toTypeHnd);
-            var_types   toType    = JITtype2varType(toJitType);
+            ClassLayout* toLayout = nullptr;
+            var_types    toType   = TypeHandleToVarType(toTypeHnd, &toLayout);
 
-            bool involvesStructType = false;
-
-            if (fromType == TYP_STRUCT)
-            {
-                involvesStructType = true;
+            assert((fromType != TYP_REF) && (toType != TYP_REF));
 
-                if (toType == TYP_STRUCT)
-                {
-                    ClassLayout* fromLayout = typGetObjLayout(fromTypeHnd);
-                    ClassLayout* toLayout   = typGetObjLayout(toTypeHnd);
-
-                    if (ClassLayout::AreCompatible(fromLayout, toLayout))
-                    {
-                        // Handle compatible struct layouts where we can simply return op1
-                        return impPopStack().val;
-                    }
-                }
-            }
-            else if (toType == TYP_STRUCT)
-            {
-                involvesStructType = true;
-            }
+            GenTree* op1 = impPopStack().val;
 
-            if (involvesStructType)
+            if ((fromTypeHnd == toTypeHnd) || (varTypeIsIntegral(fromType) && varTypeIsIntegral(toType)) ||
```

Comment:
>  `(varTypeIsIntegral(fromType) && varTypeIsIntegral(toType))`

I don't think this is correct for small types. Consider: `BitCast<short, ushort>(*pShort)` - this should clear the upper bits.